explainer.md

```diff
@@ -1,624 +1,397 @@
+
+### CSS animation
+
+CSS can be used to build on automatic animations, or override the default.
+
+```css
+::page-transition-container(header) {
+  /* … */
 }
 ```
 
-*   The prepare API initiates step 1 to 3 to save the state of shared elements
-    in the current DOM. The API takes a callback once the save operation
-    finishes executing.
-*   The async callback initiates load of the next page and initiates step 4
-    which suppresses rendering.
-*   When the callback returns, the new DOM is considered ready for first render.
-    This starts step 5 onwards to create new pseudo elements and start
-    animations.
+Element selectors:
 
-## Additional Script APIs
+- `::page-transition-container(name)` - Select the transition part containers of a given `page-transition-tag`.
+- `::page-transition-part(name)` - Select the transition parts of a given `page-transition-tag`.
+- `::page-transition-texture-outgoing(name)` - Select the outgoing texture of a given `page-transition-tag`.
+- `::page-transition-texture-incoming(name)` - Select the incoming texture of a given `page-transition-tag`.
+- `::page-transition-root-outgoing` - Select the outgoing root texture.
+- `::page-transition-root-incoming` - Select the incoming root texture.
+- `::page-transition-background` - This can be set for cases where something needs to be rendered underneath the root textures.
 
-The following example shows how developers can configure the transition in
-script for an MPA.
+These will be selecting a pseudo-element, or an element in a UA-created shadow DOM.
 
-### Old Document
+```css
+::page-transition-container(header) {
+  animation-delay: 300ms;
+}
+```
+
+CSS can be used to make changes to the automatic animation, or override `animation-name` to remove the default.
+
+### Web animation API
 
 ```js
-addEventListener("navigate", (event) => {
-  // Add sharedid attribute to elements to offer for the transition
-  // based on current document state.
-  document.querySelector(".header").sharedid="header-id";
-
-  // setData can be used to pass opaque contextual information to the
-  // new page. The argument type is |any|.
-  document.documentTransition.setData({ version: 123 });
+// In Page-B
+document.performTransition((transition) => {
+  // Build up transition parts, then…
+  transition.root.querySelector('[part=header]').animate(…);
 });
 ```
 
-### New Document
+- Open question: What's the deadline for calling `performTransition`?
+
+Again, this is a half-baked sketch. This example assumes that elements in the shadow root are given a part attribute, but there's likely some better way to address the elements.
+
+If the 'stage' of the transition is exposed as a shadow root, the developer can interact with the elements in a regular way. The developer could even create elements specifically for the transition.
```

Comment:
I was thinking you could add an element to perform some kind of fade-to-back, or a wipe effect, or even add a `<canvas>` and do some really weird shit.

What's the downside to letting devs work with this like regular DOM?

Reply:
My main tension comes from the fact that we're mirroring state in the actual DOM in this DOM representation that we're creating in the overlay/top layer. It's cached for old page but "live" for the new page. Right now it's a flat list of images but would include the hierarchy to support nested shared elements.

The more things we let developers modify in this DOM representation, the more difficult it becomes to reason about how that interacts with the browser keeping it in sync and defining how the 2 states should be merged/transitioned.

If we see legit use-cases for fancy stuff from power users, then I'd rather we just provide the primitives being used to build this representation (cached image, computed properties, style from old page and a live image for an element on the new page) directly to the developer and they can create their own overlay/DOM representation. Vlad actually also mentioned something along these lines.

For now let's try to limit the complexity to the use-cases we know. And I think hooks for animating these elements and reading computed info (ComputedStyle, getBoundingClientRect) in script is sufficient.

Reply:
The downside is that it's hard, and potentially dangerous (in the sense of accidentally very hard complexity in implementations, but also possibly security issues), to let developers do anything. For example, if we let them stick arbitrary stuff in the top layer whenever they want, then the top layer loses its value as a special place jealously guarded by the UA. Then we'd have to say how all this stuff is reflected in the DOM, mutation observers, holding onto references for elements which lost their cached backing, etc. If we expose element() to arbitrary content then they can create circularities in the rendering pipeline/tree ordering.

This is also why I'm in favor of pseudo elements instead of shadow DOM. Pseudo elements don't exist in the DOM, they only exist in the render box tree. Likewise, the top layer is a render box tree concept, not a DOM one. This has a lot of advantages, because it makes it possible for the UA to lift stuff into fullscreen without disturbing or confusing virtual DOM frameworks, event listeners, or other script stuff that operates on the DOM. Likewise, with this API, we can use the power of the render box tree and the UA implementation to elegantly lift stuff out of the DOM into the top layer, and also allow the developer to customize it (just like they can do for fullscreen and dialogs) via style sheets. And based on UA timeouts, or user overrides, we can end the animations when we want, or even skip them entirely, without the developer having to do anything (or being able to do anything, in the case of the user overriding).

If we find tons of demand for this feature for advanced modes where there are different topologies required, that's a great problem to have and we can certainly build augmented versions of this API in the future to meet those needs. In the meantime, we should go with a solution that elegantly solves the know problem and fits into existing concepts and implemented features of browsers as much as possible.